[otp] Remove cycle 0, second try

m12i opened the pull request.

danieljy commented:
Will be great to have the integration test in, and we can see the diff that this change causes!

I'll come back to this once that integration test is done.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LCnrXqJBSnsPlkG37yJ:b-iusn6e)*
<!-- Sent from Reviewable.io -->

m12i commented:
This PR is now tested with the integration test. All passed. 

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LD89Ksw-KYc5WPS2DFI:b-s9pmnq)*
<!-- Sent from Reviewable.io -->

danieljy commented:
@m12i Can you create a different PR adding a test case for this first so we see it resolved?

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LDCU95x6GIE6h-BEHwn:b-i3swvd)*
<!-- Sent from Reviewable.io -->

danieljy commented:
Also, add a test case in that PR for starting at exact location of bike.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LDDIU_d53JaffOSUVkM:b8bc65m)*
<!-- Sent from Reviewable.io -->

m12i commented:
Added test in this [PR](https://github.com/sidewalklabs/OpenTripPlanner/pull/30) for cycle 0.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LDEeV7343-2tCvFO7fS:bx1bsp6)*
<!-- Sent from Reviewable.io -->

m12i commented:
Merged the previous PR to this one, and as you can see the Cycle 0 is now gone. 
There are changes in `duration`s and `distance`s. My guess is those are cause by some java collections not guaranteeing the order of items, hence, when we link (bike/transit) stations to the graph, the change in the order of items might result in slightly different graphs. If that's the case, we may be able to fix that by serializing the graph once and load it for integration test as oppose to building the graph `--inMemory` each time.   

My suggestion is to merge this PR as is, and fix this issue in another PR, if the issue of `duration` and `distance` changes  repeats.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LDIPuL3-Qp-fBGbdNzS:b-jknhta)*
<!-- Sent from Reviewable.io -->

danieljy commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

But maybe rename those variables and add comments!

---

Reviewed 4 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, all discussions resolved.

---

*[src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java, line 254 at r1](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-LDMf1Hi6vMIrBpogTDO:-LDMf1Hj8eSzZ04RXtCW:b-4k69mm) ([raw file](https://github.com/sidewalklabs/opentripplanner/blob/626dfa8995eb830683424db03c8edadfcf41399c/src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java#L254)):*
> ```Java
> 
>         int[] legIndexPairs = {0, states.length - 1};
>         List<int[]> legsIndexes = new ArrayList<int[]>();
> ```

Wow these variable names.

Should the first one be called `legIndexPair` and the second one be called `legIndexPairs`?
The second one is a list of all the pairs, right?

---

*[src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java, line 273 at r1](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-LDMffGFBO3X76JP74lC:-LDMffGFBO3X76JP74lD:b-l5wbdf) ([raw file](https://github.com/sidewalklabs/opentripplanner/blob/626dfa8995eb830683424db03c8edadfcf41399c/src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java#L273)):*
> ```Java
>                         // Make the states[i] the start of a new leg.
>                         if (i < 2 ) {
>                             //  Note: This should never happen. The value of `i` must always be >= 2 here.
> ```

because...

---

*[src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java, line 276 at r1](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-LDMgqxTB-bUNrlEJFSS:-LDMgqxTB-bUNrlEJFST:buq8yj1) ([raw file](https://github.com/sidewalklabs/opentripplanner/blob/626dfa8995eb830683424db03c8edadfcf41399c/src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java#L276)):*
> ```Java
>                             throw new RuntimeException(format("The value of i has to be greater than 1, but i = %d", i));
>                         }
>                         legIndexPairs = new int[] {i-1, states.length - 1};
> ```

Can you add a comment here about why you look one back now when setting up the new candidate pair (effectively ignoring the entire leg switch leg) while that's not necessary for a transit leg switch, for example? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LDMj77f7A90fnQGDB9k:b3pntnz)*
<!-- Sent from Reviewable.io -->

m12i commented:
I believe we need to introduce a new traverse mode. LEG_SWITCH shouldn't be used for links and besides this case is the first time we need to postpone the decision of setting the traverse mode. So it makes sense to add a new traverse mode. 

Let me know what you think

---

Review status: 3 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java, line 254 at r1](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-LDMf1Hi6vMIrBpogTDO:-LDMnuOz4aTPPrUWAFph:bv9z8l5) ([raw file](https://github.com/sidewalklabs/opentripplanner/blob/626dfa8995eb830683424db03c8edadfcf41399c/src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java#L254)):*
<details><summary><i>Previously, danieljy (Daniel Yehuda) wrote…</i></summary><blockquote>

Wow these variable names.

Should the first one be called `legIndexPair` and the second one be called `legIndexPairs`?
The second one is a list of all the pairs, right?
</blockquote></details>

I am not happy with the variable names in OTP, but I am not sure it's a good idea to change them at the moment.

---

*[src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java, line 273 at r1](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-LDMffGFBO3X76JP74lC:-LDNfYFU-csKh7pBCtND:b-9unw5o) ([raw file](https://github.com/sidewalklabs/opentripplanner/blob/626dfa8995eb830683424db03c8edadfcf41399c/src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java#L273)):*
<details><summary><i>Previously, danieljy (Daniel Yehuda) wrote…</i></summary><blockquote>

because...
</blockquote></details>

deleted

---

*[src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java, line 276 at r1](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-LDMgqxTB-bUNrlEJFSS:-LDNf_WjB_5nsWogoO-8:b-n9tkhd) ([raw file](https://github.com/sidewalklabs/opentripplanner/blob/626dfa8995eb830683424db03c8edadfcf41399c/src/main/java/org/opentripplanner/api/resource/GraphPathToTripPlanConverter.java#L276)):*
<details><summary><i>Previously, danieljy (Daniel Yehuda) wrote…</i></summary><blockquote>

Can you add a comment here about why you look one back now when setting up the new candidate pair (effectively ignoring the entire leg switch leg) while that's not necessary for a transit leg switch, for example? 
</blockquote></details>

changed

---


*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LDNfbcV34DmmzmH8jkU:bkgi17n)*
<!-- Sent from Reviewable.io -->

m12i commented:
ping

---

Review status: 3 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sidewalklabs/opentripplanner/25#-:-LDgMiyn4oAZw8MupvYi:bb7mm0n)*
<!-- Sent from Reviewable.io -->
